### PR TITLE
Altair message-id topic

### PR DESCRIPTION
**Pull request by Menduist**
cc @zah

### CODE REVIEW

File: beacon_chain/networking/eth2_network.nim [hunk 1]
```diff
@@ -1500,25 +1500,44 @@ proc getPersistentNetKeys*(rng: var BrHmacDrbgContext,
       pubKey = privKey.getKey().expect("working public key from random")
     NetKeyPair(seckey: privKey, pubkey: pubKey)
 
-func gossipId(data: openArray[byte], valid: bool): seq[byte] =
+
+func gossipId(data: openArray[byte], topic: string, valid: bool): seq[byte] =
   # https://github.com/ethereum/eth2.0-specs/blob/v1.0.1/specs/phase0/p2p-interface.md#topics-and-messages
+  # and
+  # https://github.com/ethereum/eth2.0-specs/blob/dev/specs/altair/p2p-interface.md#topics-and-messages
```
Comment (tersec) [inline beacon_chain/networking/eth2_network.nim]: It's useful, to track which parts of the implementation need to be checked when updating to new spec versions, to link to the specific version it was based on, e.g., https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.7/specs/altair/p2p-interface.md#topics-and-messages

File: beacon_chain/networking/eth2_network.nim [hunk 2]
```diff
@@ -1500,25 +1500,44 @@ proc getPersistentNetKeys*(rng: var BrHmacDrbgContext,
...
 
-  result = newSeq[byte](20)
-  result[0..19] = messageDigest.data.toOpenArray(0, 19)
+  return messageDigest.data[0..19].toSeq()
+
+func isAltairTopic(topic: string): bool =
+  #TODO make a better one one when we have global constants
+  topic.startsWith("sync_committee_") or
+  topic.startsWith("sync_committee_contribution_and_proof") or
+  topic.startsWith("beacon_block")
```
Comment (tersec) [inline beacon_chain/networking/eth2_network.nim]: It looks like:
> Note that the `ForkDigestValue` path segment of the topic separates the old and the new `beacon_block` topics.

Thus, not all `beacon_block`-prefixed topics are Altair.

Comment (Menduist) [reply beacon_chain/networking/eth2_network.nim]: Oh yeah, just checked the topic creation spec, this function is entirely wrong. I'll update it

**Comment by Menduist** [comment]
Ok so the new version only uses the `ForkDigestValue` to find Altair topics, though I don't know how should I get Altair's `ForkDigestValue`?
I guess the system to track the `ForkDigestValue` of both Phase0 & Altair during the fork is not setup yet, so I can't do it ATM?

**Comment by tersec** [comment]
> Ok so the new version only uses the `ForkDigestValue` to find Altair topics, though I don't know how should I get Altair's `ForkDigestValue`?
> I guess the system to track the `ForkDigestValue` of both Phase0 & Altair during the fork is not setup yet, so I can't do it ATM?

The `ForkDigestValue` isn't just a function of Altair and phase 0.

https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.7/specs/phase0/p2p-interface.md#topics-and-messages describes it:

- `ForkDigestValue` - the lowercase hex-encoded (no "0x" prefix) bytes of `compute_fork_digest(current_fork_version, genesis_validators_root)` where
    - `current_fork_version` is the fork version of the epoch of the message to be sent on the topic
    - `genesis_validators_root` is the static `Root` found in `state.genesis_validators_root`

In particular, the `current_fork_version` effectively distinguishes between phase 0 and Altair. However, https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.7/specs/phase0/beacon-chain.md#compute_fork_data_root shows how that `ForkDigestValue` is effectively downstream of:
```python
    return hash_tree_root(ForkData(
        current_version=current_version,
        genesis_validators_root=genesis_validators_root,
    ))
```
i.e. an opaque value in itself.

You're correct: currently the `unstable` branch (and none of my current branches in development) doesn't address this point.

It's possible, but inefficient given the sheer volume of messages across which these values don't change, to recompute each time based on `current_version` and `genesis_validators_root`, both of which are in both phase 0 and Altair `BeaconState`s (https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.7/specs/phase0/beacon-chain.md#beaconstate and https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.7/specs/altair/beacon-chain.md#beaconstate).

Currently, Nimbus tracks this via the `ForkDigest` type, part of the beacon node:
https://github.com/status-im/nimbus-eth2/blob/41e0a7abc07cbb60afdc2871ff5fcb21a4ae5e4e/beacon_chain/beacon_node_common.nim#L52

Any given Nimbus node is assumed to be on only one network at a time, which implies an invariant `genesis_validators_root`, but the `current_fork_version` is now somewhat variable. One approach is to expand that `BeaconNode.forkDigest` field to hold up to as many forks' worth of digests as Nimbus currently supports -- then one can check against them for this sort of purpose.

Presumably @zah has already encountered this issue, as part of broader networking issues, and might already have an approach here.

**Comment by zah** [comment]
Indeed, I've finished the implementation in my [own PR](https://github.com/status-im/nimbus-eth2/pull/2676) where the `ForkDigest` values are cached and shared between several components such as the DAG and the networking.

**Pull request merged**